crates/feldera-storage/src/backend/mod.rs

```diff
@@ -0,0 +1,244 @@
+///
+/// Note: If we decide to e.g., stick with only monoio we can avoid this
+/// abstraction.
+pub struct WriteBuffer {
+    inner: InnerWriteBuffer,
+}
+
+impl Into<WriteBuffer> for Vec<u8> {
+    fn into(self) -> WriteBuffer {
+        WriteBuffer {
+            inner: InnerWriteBuffer::Vec(self),
+        }
+    }
+}
+
+impl Deref for WriteBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        match self.inner {
+            #[cfg(feature = "glommio")]
+            InnerWriteBuffer::Glommio(ref x) => x.as_bytes(),
+            InnerWriteBuffer::Vec(ref x) => x.as_slice(),
+        }
+    }
+}
+
+impl DerefMut for WriteBuffer {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self.inner {
+            #[cfg(feature = "glommio")]
+            InnerWriteBuffer::Glommio(ref mut x) => x.as_bytes_mut(),
+            InnerWriteBuffer::Vec(ref mut x) => x.as_mut_slice(),
+        }
+    }
+}
+
+/// A buffer suitable reading a file using Direct I/O over io_uring.
+pub enum ReadBuffer {
+    #[cfg(feature = "glommio")]
+    Glommio(ReadResult),
+    Vec(Vec<u8>),
+}
+
+impl Deref for ReadBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            #[cfg(feature = "glommio")]
+            ReadBuffer::Glommio(x) => x.as_ref(),
+            ReadBuffer::Vec(x) => x.as_slice(),
+        }
+    }
+}
+
+/// An error that can occur when using the storage backend.
+#[derive(Error, Debug)]
+pub enum StorageError {
+    #[cfg(feature = "glommio")]
+    #[error("Got IO error during glommio operation")]
+    Io(#[from] glommio::GlommioError<()>),
+    #[error("Got IO error during monoio operation")]
+    StdIo(#[from] std::io::Error),
+}
+
+/// A trait for a storage backend to implement so client can create/delete
+/// files.
+pub trait StorageControl {
+    /// Creates a new persistent file used for writing data.
+    ///
+    /// Returns a file-descriptor that can be used for writing data.
+    /// Note that it is not possible to read from this file until
+    /// [`StorageWrite::complete`] is called and the [`FileHandle`] is
+    /// converted to an [`ImmutableFileHandle`].
+    async fn create(&self) -> Result<FileHandle, StorageError>;
+
+    /// Deletes a previously completed file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for reading.
+    async fn delete(&self, fd: ImmutableFileHandle) -> Result<(), StorageError>;
+
+    /// Deletes a previously created file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for writing.
+    ///
+    /// Use [`delete`] for deleting a file that has been completed.
+    async fn delete_mut(&self, fd: FileHandle) -> Result<(), StorageError>;
+}
+
+/// A trait for a storage backend to implement so clients can write to files.
+pub trait StorageWrite {
```

Comment:
I'm a little surprised that the `fd` is an argument to the functions in this trait, instead of being owned by the object.

Reply:
Note that the fd is not the `OS-level fd` but instead an fd that the buffer manager hands out. Maybe I should rename it to file-handle to make it clearer?